src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/StateSynchronizationObserver.cs

```diff
@@ -256,6 +438,132 @@ private void ResetAssetCaches()
+            switch (AssetState.Status)
+            {
+                case AssetStateStatus.None:
+                    if (assetCacheCount > 0)
+                    {
+                        AssetState = new AssetState { Status = AssetStateStatus.Preloaded };
+                    }
+                    break;
+
+                case AssetStateStatus.Preloaded:
+                    if (assetCacheCount <= 0)
+                    {
+                        AssetState = new AssetState { Status = AssetStateStatus.None };
+                    }
+                    break;
+
+                case AssetStateStatus.RequestingAssetBundle:
+                case AssetStateStatus.DownloadingAssetBundle:
+                case AssetStateStatus.AssetBundleLoaded:
+                case AssetStateStatus.NonePreloadedAndNoAssetBundleAvailable:
+                case AssetStateStatus.ErrorDownloadingAssetBundle:
+                case AssetStateStatus.ErrorLoadingAssetBundle:
+                    // No adjustments needed
+                    break;
+
+                case AssetStateStatus.Unknown:
+                default:
+                    Debug.LogError($"Unexpected asset state status \"{AssetState.Status}\".", this);
+                    break;
+            }
+        }
+
+        public static string FormatByteProgress(int bytesSoFar, int totalBytes)
+        {
+            GetBytesFormat(totalBytes, out float bytesDivisor, out string bytesFormat, out string bytesUnitSpecifier);
+
+            var bytesSoFarText = (bytesSoFar / bytesDivisor).ToString(bytesFormat);
+            var totalBytesText = (totalBytes / bytesDivisor).ToString(bytesFormat);
+
+            var percentCompleteText = (100.0 * bytesSoFar / totalBytes).ToString("N1");
+
+            return $"{percentCompleteText}% ({bytesSoFarText}/{totalBytesText} {bytesUnitSpecifier})";
+        }
+
+        public static string FormatBytes(int totalBytes)
+        {
+            GetBytesFormat(totalBytes, out float divisor, out string format, out string unitSpecifier);
+
+            return $"{(totalBytes / divisor).ToString(format)} {unitSpecifier}";
         }
+
+        private static void GetBytesFormat(int totalBytes, out float divisor, out string format, out string unitSpecifier)
+        {
+            if (totalBytes >= 1024 * 1024)
+            {
+                divisor = 1024 * 1024;
+                format = "N1";
+                unitSpecifier = "MB";
+            }
+            else if (totalBytes >= 1024)
+            {
+                divisor = 1024;
+                format = "N1";
+                unitSpecifier = "KB";
+            }
+            else
+            {
+                divisor = 1;
+                format = "N0";
+                unitSpecifier = "B";
+            }
+        }
+
+        private class AssetBundleReceive
+        {
+            public byte[] Data;
+            public int NextDataToReceiveIndex;
+        }
+    }
+
+    public enum AssetStateStatus
+    {

```

Comment:
nit: may be worth making this an internal/private enum declaration. Otherwise we need summaries for each value